File: 1on1/new_course/1320minimumdistance.py
class Solution:
    def minimumDistance(self, word: str) -> int:
        # build the board   0~26   x=6, y=5
        chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        B = collections.defaultdict(lambda: [])
        for k in range(26):
            (d, r) = divmod(k, 6)
            B[chars[k]] = [d, r]
        res = math.inf

        # divide the board
        def dfs(cur_dis, prev_l, prev_r, idx, word):
            nonlocal res
            if idx >= len(word):
                res = min(res, cur_dis)
                return
            elif cur_dis >= res:  # optimize
                return
            # try left-finger
            if prev_l:
                tmp_dis = abs(B[word[idx]][0]-B[prev_l][0]) + \
                    abs(B[word[idx]][1]-B[prev_l][1])
                if tmp_dis < 5:  # optimize
                    dfs(cur_dis+tmp_dis, word[idx], prev_r, idx+1, word)
            else:
                dfs(cur_dis, word[idx], prev_r, idx+1, word)
            # try right-finger
            if prev_r:
                tmp_dis = abs(B[word[idx]][0]-B[prev_r][0]) + \
                    abs(B[word[idx]][1]-B[prev_r][1])
                if tmp_dis < 5:  # optimize
                    dfs(cur_dis+tmp_dis, prev_l, word[idx], idx+1, word)
            else:
                dfs(cur_dis, prev_l, word[idx], idx+1, word)

        dfs(0, "", "", 0, word)

        return res



import math, collections

class Solution:
    def minimumDistance(self, W: str) -> int:
        def d(a, b):
            return a and abs(a // 6 - b // 6) + abs(a % 6 - b % 6)

        #dp, dp2 = {(0, 0): 0}, collections.defaultdict(lambda:3000)
        dp, dp2 = collections.defaultdict(lambda:0), collections.defaultdict(lambda:3000)
        dp[(0,0)] = 0
        for c in (ord(c) + 1 for c in W):
            #print("c:", c)
            for a, b in dp:
                # dp2[(c, b)] = min(dp2.get((c, b), 3000), dp[a, b] + d(a, c))
                # dp2[(a, c)] = min(dp2.get((a, c), 3000), dp[a, b] + d(b, c))
                #print("a:", a, "b:", b, "c:", c)
                dp2[(c, b)] = min(dp2[(c, b)], dp[(a, b)] + d(a, c))
                dp2[(a, c)] = min(dp2[(a, c)], dp[(a, b)] + d(b, c))
            dp, dp2 = dp2, collections.defaultdict(lambda:3000)
        return min(dp.values())

File: 1on1/new_course/1320MinimumDistance.java
class Solution {
    public int minimumDistance(String word) {
        String chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        HashMap<Character, Pos> board = new HashMap<>();
        for (int i = 0; i < chars.length(); i++) {
            Integer dvd = i / 6;
            Integer rem = i % 6;
            Pos curPos = new Pos(dvd, rem);
            board.put(chars.charAt(i), Pos);
        }
        System.out.println("board:" + board.toString());

        Status start = new Status();
        HashMap<Status, Integer> dp = new HashSet<>();
        dp.put(start, 0);

        for (Character c : word.toCharArray()) {
            HashSet<Status> nxtdp = new HashSet<>();
            for (Status status : dp) {
                // for left-finger
                nxtStatusl = new Status(status.left, status.right);
                nxtStatusl.left = c;
                Integer disl = getDis(status.left, c, board);
                if (nxtdp.get(nxtStatusl) != null) {
                    disl = Math.min(nxtdp.get(nxtStatusl), disl);
                }
                nxtdp.put(nxtStatusl, disl);
                // for right-finger
                nxtStatusr = new Status(status.left, status.right);
                nxtStatus.right = c;
                Integer disr = getDis(status.right, c, board);
                if (nxtdp.get(nxtStatus) != null) {
                    disr = Math.min(nxtdp.get(nxtStatus), disr);
                }
                nxtdp.put(nxtStatus, disr);
            }
            dp = nxtdp;
        }
        return Math.min(dp.values());
    }

    private Integer getDis(Character c1, Character c2, HashMap board) {
        if (c1 == null || c2 == null)
            return 0;
        Pos pos1 = board.get(c1);
        Pos pos2 = board.get(c2);
        return abs(pos1.row - pos2.row) + abs(pos1.col - pos2.col);
    }
}

class Pos {
    Integer row;
    Integer col;

    public void Pos(Integer r, Integer c) {
        this.row = r;
        this.col = c;
    }
}

class Status {
    Character left;
    Character right;

    public void Status() {
    }

    public void Status(Character l, Character r) {
        this.left = l;
        this.right = r;
    }
}